### PR TITLE
Add support for Oracle GraalVM

### PR DESCRIPTION
**Pull request by fniephaus**
**Description:**
This PR adds Oracle GraalVM JDK to the list of supported distributions.

**Related issue:** https://github.com/actions/setup-java/pull/401

**Check list:**
- [x] Mark if documentation changes are required.
- [x] Mark if tests were added or updated to cover the changes.

### CODE REVIEW

**Comment by Pogie10** [comment]
PAG TATAMPALIN KO MGA MUKA NYO EE 👊👊👊

**Review by tisonkun — APPROVED** [review]
Cool!

Out thread - Does this action support OpenJDK?

**Comment by fniephaus** [comment]
Any update on this?

**Comment by wetneb** [comment]
@IvanZosimov if you could find time to merge this, it would be really amazing! Thank you for your work.

**Comment by Yaytay** [comment]
Ideally this should support the new release and  approach to naming from GraalVM.

**Comment by fniephaus** [comment]
@IvanZosimov and @dmitry-shibanov, do you have an update on this? There's now a merge conflict that needs to be resolved but I don't want to do this work if a review and approval still take time.

**Comment by fniephaus** [comment]
I'm assuming this is still pending licensing approval, @IvanZosimov? Please let me know if there's anything we can help with.

**Comment by fniephaus** [comment]
Maybe @brunoborges could help move this forward?

**Comment by fniephaus** [comment]
@aparnajyothi-y I see that [SapMachine support is planned for Q4](https://github.com/actions/setup-java/pull/614#issuecomment-2244917425). Can we please proceed with this and add Oracle GraalVM support, too?

**Comment by aparnajyothi-y** [comment]
Hello @fniephaus, We are working to prioritize this feature. We will update you once we have some feedback :)

**Comment by fniephaus** [comment]
Thanks, @aparnajyothi-y. Let me know when this is planned so I can help resolve the merge conflicts. :)

**Comment by SandraAhlgrimm** [comment]
Hi @aparnajyothi-y, any updates here?

**Comment by HarithaVattikuti** [comment]
Hello Everyone, we are in the process of reviewing this PR. 
Feedback will be provided once the review is complete.

**Comment by fniephaus** [comment]
> Feedback will be provided once the review is complete.

Thanks! Let me know and we can work on resolving merge conflicts.

**Comment by mahabaleshwars** [comment]
Hi @fniephaus, could you please resolve the conflicts so we can proceed with testing and approving the PR?

**Comment by fniephaus** [comment]
Hi @mahabaleshwars,
I have resolved the merge conflicts and have added support for GraalVM EA builds. Please test and review the PR and let me know if anything needs changing.

**Review by alina-yur — APPROVED** [review]

**Review by mahabaleshwars — APPROVED** [review]

**Comment by fniephaus** [comment]
I see there are new merge conflicts. When will this PR be merged? I wouldn't want to keep resolving merge conflicts.

**Review by priya-kinthali — APPROVED** [review]

**Review by aparnajyothi-y — APPROVED** [review]

**Comment by fniephaus** [comment]
I have resolved the merge conflicts again. It'd be great if this could be merged soon, thanks!

**Comment by fniephaus** [comment]
The failure is due to hitting the API rate limit for some reason: https://github.com/actions/setup-java/actions/runs/10806059192/job/29974203938#step:8:1235

**Comment by fniephaus** [comment]
All tests are passing and there are plenty of approvals. When will this be merged? :)

File: README.md [hunk 1]
```diff
@@ -109,6 +109,7 @@ Currently, the following distributions are supported:
 | `oracle` | Oracle JDK | [Link](https://www.oracle.com/java/technologies/downloads/) | [Link](https://java.com/freeuselicense)
 | `dragonwell` | Alibaba Dragonwell JDK | [Link](https://dragonwell-jdk.io/) | [Link](https://www.aliyun.com/product/dragonwell/)
 | `sapmachine` | SAP SapMachine JDK/JRE | [Link](https://sapmachine.io/) | [Link](https://github.com/SAP/SapMachine/blob/sapmachine/LICENSE)
+| `graalvm` | Oracle GraalVM | [Link](https://www.graalvm.org/) | [Link](https://www.oracle.com/downloads/licenses/graal-free-license.html)
```
Comment (Marcono1234) [inline README.md:112]: This license information might be incorrect or at least misleading. According to https://www.oracle.com/java/technologies/javase/jdk-faqs.html#GraalVM-licensing (section "What is the licensing for currently available Oracle GraalVM Releases?" and "How long will Oracle use the GFTC and what happens afterwards?") the GraalVM Free Terms and Conditions (GFTC) only apply for one year after the subsequent LTS, so if I understand it correctly
- Oracle GraalVM 17 licensing changed September 2024
- Oracle GraalVM 21 licensing will change September 2026

What exactly does this mean if you use `graalvm` in your workflow?
- Are you still using GraalVM under the GFTC after this time but simply don't receive any updates,
- or are you unwittingly violating the Oracle license?

Related: https://github.com/graalvm/setup-graalvm/issues/107

Comment (fniephaus) [reply README.md:112]: > This license information might be incorrect or at least misleading.

[Script Friendly URLs](https://www.oracle.com/java/technologies/jdk-script-friendly-urls/) are used to pull Oracle GraalVM with `setup-java` and they only provide Oracle GraalVM under the GraalVM Free Terms and Conditions license. So I don't see why the license information would be incorrect.

> What exactly does this mean if you use graalvm in your workflow after these dates?

The action uses the [latest URLs to resolve `java-version: '17'`](https://github.com/actions/setup-java/blob/83a06ff9d9aa70f76a8d73278e646c20b2bf1ae5/src/distributions/graalvm/installer.ts#L85) for Oracle GraalVM and those URLs are no longer available. In practice, this means that `java-version: '17'` no longer works, but you can use `java-version: '17.0.12'` to [pull from the archive](https://github.com/actions/setup-java/blob/83a06ff9d9aa70f76a8d73278e646c20b2bf1ae5/src/distributions/graalvm/installer.ts#L82). By the way, new CPU releases of Oracle GraalVM for JDK 17 are available [here](https://www.oracle.com/java/technologies/downloads/?er=221886#graalvmjava17).

Comment (Marcono1234) [reply README.md:112]: > So I don't see why the license information would be incorrect.

I had assumed the license automatically switches from GFTC to the Oracle license, but as you have explained that is not the case. Thanks for the clarification!

**Comment by mahabaleshwars** [comment]
Hi @fniephaus, I have created PR #849 to address the API rate limit issue caused by the test cases in GraalVM. Please review the changes when you have a moment. Let me know if you have any questions or feedback. Thank you!

**Pull request merged**